Lib/test/test_inspect/test_inspect.py

```diff
@@ -4623,6 +4657,42 @@ class D2(D1):
         self.assertEqual(inspect.signature(D2), inspect.signature(D1))
 
 
+class TestSignatureFromFrame(unittest.TestCase):
+    def test_signature_from_frame(self):
+        def inner(a=1, /, b=2, *e, c: int = 3, d, **f) -> None:
+            global fr
```

Comment:
Using a global sounds like a bad idea. You can use a  "nonlocal" instead.

Sometimes, I use a mutable type instead, which is more or less the same:

```
ns = {}
def func():
    ns['name'] = value
```